cf-agent/package_module.c

```diff
@@ -0,0 +1,1604 @@
+{
+    /* Check if we are not having 'latest' version. */
+    if (policy_data->package_version &&
+            StringSafeEqual(policy_data->package_version, "latest"))
+    {
+        Log(LOG_LEVEL_ERR, "Package version 'latest' not supported for"
+                "absent package promise");
+        return PROMISE_RESULT_FAIL;
+    }
+    
+    /* Check if package exists in cache */
+    int is_in_cache = IsPackageInCache(ctx, wrapper, package_name,
+                                       policy_data->package_version,
+                                       policy_data->package_architecture);
+    if (is_in_cache == 1)
+    {
+        /* Remove package(s) */
+        PromiseResult res;
+        
+        if (action == cfa_warn || DONTDO)
+        {
+            Log(LOG_LEVEL_NOTICE, "Need to remove package: %s", package_name);
+            res = PROMISE_RESULT_FAIL;
+        }
+        else
+        {
+            res = RemovePackage(package_name,
+                    policy_data->package_options, policy_data->package_version,
+                    policy_data->package_architecture, wrapper);
+
+            if (res == PROMISE_RESULT_CHANGE)
+            {
+                /* Check if package was removed. */
+                return ValidateChangedPackage(policy_data, wrapper, 
+                        &((PackageInfo){.name = package_name, 
+                                        .version = policy_data->package_version, 
+                                        .arch = policy_data->package_architecture}),
+                        NEW_PACKAGE_ACTION_ABSENT);
+            }
+        }
+        
+        return res;
+    }
+    else if (is_in_cache == -1)
+    {
+        Log(LOG_LEVEL_INFO, "Error occurred while checking package '%s' "
+            "existence in cache.", package_name);
+        return PROMISE_RESULT_FAIL;
+    }
+    else
+    {
+        /* Package is not in cache which means it is already removed. */
+        Log(LOG_LEVEL_DEBUG, "Package '%s' not installed. Skipping removing.",
+                package_name);
+        return PROMISE_RESULT_NOOP;
+    }
+}
+
+
+/* IMPORTANT: This must be called under protection of 
+ * GLOBAL_PACKAGE_PROMISE_LOCK_NAME lock! */
+bool UpdateSinglePackageModuleCache(EvalContext *ctx,
+                                    const PackageModuleWrapper *module_wrapper,
+                                    UpdateType type, bool force_update)
+{
+    assert(module_wrapper->package_module->name);
+    
+    if (!force_update)
+    {
+        if (module_wrapper->package_module->installed_ifelapsed == CF_NOINT ||
+            module_wrapper->package_module->updates_ifelapsed == CF_NOINT)
+        {
+            Log(LOG_LEVEL_ERR, "Package module body constraints error: %s %d %d",
+                module_wrapper->package_module->name, 
+                module_wrapper->package_module->installed_ifelapsed,
+                module_wrapper->package_module->updates_ifelapsed);
```

Comment:
A bit cryptic. I would say ``Not all attributes specified in package body '%s': query_installed_ifelapsed = %d, query_updates_ifelapsed = %d``.